### DISCUSSION TITLE
Need Advice for improve performance

### DISCUSSION THREAD

**Discussion by slookin** [lark-parser/lark#1116] [General]
I have problem with performance using Lark.

[https://github.com/slookin/stuff/tree/master/lark_demo_speed](https://github.com/slookin/stuff/tree/master/lark_demo_speed)

There are EBNF notation which parse plantuml sequence diagram, but it takes minutes for just 300 lines of plant uml.

I know about parser = larl, but I'm not able to prepare grammar without collisions. (may be I'm not so strong in EBNF or platuml language is not so good).

Do you have any ideas/advices how improve performance? (I need parse regularly 40-50 such diagrams, and right now it takes hour(s)).

Thanks in advance.

**Comment by erezsh** [comment]
First of all, change the file extension from `.ebnf` to `.lark` so I can read it with syntax highlighting.

But quick guess, your extensive use of `/(.)+/ ` probably causes a lot of over-scanning to the end of the file. Better to include all the characters you actually need to match (say, `\w`). And if you really need a catch-all, better to use lazy form (`.+?`)

**Reply by slookin** [reply]
extension changed.

**Reply by slookin** [reply]
I'm trying to use lazy form for regexp, but it doesn't work as expects.
for example if change 
line 7 from 
_start: /\d+/ 
to 
_start: /\d+?/
then Lark not able to parse first line in example.

**Reply by erezsh** [reply]
Lazy means it will take the minimum amount. Meaning if it has 999, it will only match the first 9. If you use lazy regexps, you need to give them some terminating condition.
